### PR TITLE
CREATE SHARDING TABLE RULE

### PR DESCRIPTION
**Pull request by isHuangXin**
CREATE SHARDING TABLE RULE  #17939.

In this PR, I finished parsing the CREATE SHARDING TABLE RULE in the yaml file and generating the corresponding distsql statement.
The solution for this PR is to first create the sharding algorithm, and then specify the sharding algorithm to create standard sharding rules

**Finally, the result of running my code is**

- sharding.yaml file content
image
image

**CREATE SHARDING TABLE RULE & CREATE SHARDING BINDING TABLE RULE**

**STANDARD mode**
- The converted distsql output 
image

- Running the converted distsql in SS proxy
image

**COMPLEX mode**
- The converted distsql output
image

- Running the converted distsql in SS proxy
image

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/constant/DistSQLScriptConstants.java [hunk 1]
```diff
@@ -63,4 +63,31 @@ public final class DistSQLScriptConstants {
...
+
+    public static final String SHARDING_TABLE = " %s ("
+            + System.lineSeparator()
+            + "DATANODES(\"%s\"),"
+            + System.lineSeparator()
+            + "%s"
+            + System.lineSeparator()
+            + ")";
+
+    public static final String DATABASE_SHARDING_STRATEGY = "DATABASE_STRATEGY(TYPE=standard, SHARDING_COLUMN=%s, SHARDING_ALGORITHM=%s),"
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/constant/DistSQLScriptConstants.java:85]: TYPE is not fixed as `standard`, please get it from configuration.

Comment (RaigorJiang) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/constant/DistSQLScriptConstants.java:85]: Completing this feature requires you to be familiar with YAML configuration:
https://shardingsphere.apache.org/document/current/en/user-manual/shardingsphere-jdbc/yaml-config/rules/sharding/

Note that when TYPE is `COMPLEX`, the key is no longer `shardingColumn`, but `shardingColumns`.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 2]
```diff
@@ -135,4 +144,94 @@ private void appendProperties(final Map<String, Object> properties, final String
...
+        return result.toString();
+    }
+
+    private void appendShardingTableRules(final ShardingRuleConfiguration shardingRuleConfig, final StringBuilder stringBuilder) {
+        stringBuilder.append(DistSQLScriptConstants.CREATE_SHARDING_TABLE);
+        Iterator<ShardingTableRuleConfiguration> iterator = shardingRuleConfig.getTables().iterator();
+        while (iterator.hasNext()) {
+            ShardingTableRuleConfiguration entry = iterator.next();
+            String tableName = entry.getLogicTable();
+            String dataNodes = entry.getActualDataNodes();
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:195]: If autoTable, here is `RESOURCES`, not `DATANODES`

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 3]
```diff
@@ -135,4 +144,94 @@ private void appendProperties(final Map<String, Object> properties, final String
...
+            if (iterator.hasNext()) {
+                result.append(DistSQLScriptConstants.COMMA);
+            }
+        }
+        return result.toString();
+    }
+
+    private void appendShardingTableRules(final ShardingRuleConfiguration shardingRuleConfig, final StringBuilder stringBuilder) {
+        stringBuilder.append(DistSQLScriptConstants.CREATE_SHARDING_TABLE);
+        Iterator<ShardingTableRuleConfiguration> iterator = shardingRuleConfig.getTables().iterator();
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:191]: Also need to consider autoTables.   shardingRuleConfig.getAutoTables()

**Review by RaigorJiang — CHANGES_REQUESTED** [review]
This PR has made great progress, but the scenarios covered are not complete, please use more types of configurations to test.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/constant/DistSQLScriptConstants.java (hunk 1 above)
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/constant/DistSQLScriptConstants.java:85]: > TYPE is not fixed as `standard`, please get it from configuration.

Fixed !

**Review by RaigorJiang — APPROVED** [review]

**Comment by RaigorJiang** [comment]
@isHuangXin Please add test cases later.

**Comment by isHuangXin** [comment]
> @isHuangXin Please add test cases later.

Get it, but I plan to add test after implementing all functions.

**Comment by RaigorJiang** [comment]
> Get it, but I plan to add test after implementing all functions.

Your PR has decreased coverage a lot, and if continues like this, future PRs may not be merged.

**Comment by isHuangXin** [comment]
> > Get it, but I plan to add test after implementing all functions.
> 
> Your PR has decreased coverage a lot, and if continues like this, future PRs may not be merged.

OK, I will finish the test unit in the next PR.

**Pull request merged**